firebase-firestore/src/main/java/com/google/firebase/firestore/model/protovalue/ObjectValue.java

```diff
@@ -0,0 +1,266 @@
+      // subdivided during each iteration. The iteration stops when the slice becomes empty.
+      SortedMap<FieldPath, Value> currentSlice =
+          currentPath.isEmpty()
+              ? overlayMap
+              : overlayMap.subMap(currentPath, createSuccessor(currentPath));
+
+      boolean modified = false;
+
+      while (!currentSlice.isEmpty()) {
+        FieldPath fieldPath = currentSlice.firstKey();
+
+        if (fieldPath.length() == currentPath.length() + 1) {
+          // The key in the slice is a leaf node. We can apply the value directly.
+          String fieldName = fieldPath.getLastSegment();
+          Value overlayValue = overlayMap.get(fieldPath);
+          if (overlayValue != null) {
+            resultAtPath.putFields(fieldName, overlayValue);
+            modified = true;
+          } else if (resultAtPath.containsFields(fieldName)) {
+            resultAtPath.removeFields(fieldName);
+            modified = true;
+          }
+        } else {
+          // Since we need a MapValue.Builder at each nesting level (e.g. to create the field for
+          // `a.b.c` we need to create a MapValue.Builder for `a` as well as `a.b`), we invoke
+          // applyOverlay() recursively with the next nesting level.
+          FieldPath nextSliceStart = fieldPath.keepFirst(currentPath.length() + 1);
+          @Nullable FieldValue existingValue = baseObject.get(nextSliceStart);
+          MapValue.Builder nextSliceBuilder =
+              existingValue instanceof ObjectValue
+                  // If there is already data at the current path, base our modifications on top
+                  // of the existing data.
+                  ? ((ObjectValue) existingValue).internalValue.getMapValue().toBuilder()
+                  : MapValue.newBuilder();
+          modified = applyOverlay(nextSliceStart, nextSliceBuilder) || modified;
+          if (modified) {
+            // Only apply the result if a field has been modified. This avoids adding an empty
+            // map entry for deletes of non-existing fields.
+            resultAtPath.putFields(
+                nextSliceStart.getLastSegment(),
+                Value.newBuilder().setMapValue(nextSliceBuilder).build());
+          }
+        }
+
+        // Shrink the subtree to contain only values after the current field path. Note that we are
+        // still bound by the subtree created at the initial method invocation. The current loop
+        // exits when the subtree becomes empty.
+        currentSlice = currentSlice.tailMap(createSuccessor(fieldPath));
+      }
+
+      return modified;
+    }
+
+    /** Create the first field path that is not part of the subtree created by `currentPath`. */
+    private FieldPath createSuccessor(FieldPath currentPath) {
+      hardAssert(!currentPath.isEmpty(), "Can't create a successor for an empty path");
+      return currentPath.popLast().append(currentPath.getLastSegment() + '0');
```

Comment:
This is also no longer needed (yay for simpler code).

I chose `'0'` because it is the first ASCII code that can be part of a valid field path. `'\0'` may trigger validation errors in our field value handling.